feat: add support for implicit return

amchiclet opened the pull request:
Partially helps with #8408 (but doesn't close it).

In languages such as Julia, Ruby, and Rust, an expression that is last executed in a function is implicitly that function's return value.

This PR supports matching these implicit returns.

For example, the pattern `return 0` can match functions such as
  ```julia
  f() = 0
  ```
and
  ```julia
  function f()
    0
  end
  ```
and more.

This feature can be disabled through the `implicit_return` rule option.

Tested with `make core-test` and added new tests for Julia, Ruby, and Rust.

Implementation:

To find nodes that may exit, we walk back from the exit until we find the first Instr node along each path.

Then the expression on the RHS of that instruction is marked as implicitly returned. The node-marking is done during the resolve process in Naming_AST. Then at CFG creation time, these marked nodes will be converted to return nodes.

amchiclet committed:
feat: add support for implicit return

Partially helps with #8408 (but doesn't close it).

In languages such as Julia, Ruby, and Rust, an expression that is
last executed in a function is implicitly that function's return value.

This PR supports matching these implicit returns.

For example, the pattern `return 0` can match functions such as
  ```julia
  f() = 0
  ```
and
  ```julia
  function f()
    0
  end
  ```
and more.

Tested with `make core-test` and added new tests for Julia, Ruby, and Rust.

IagoAbal commented on tests/patterns/rust/implicit_return.rs:
```diff
@@ -0,0 +1,61 @@
+fn f() {
+  // ERROR:
+  let g = || "implicit return for lambda";
```
What does this mean in Rust?

amchiclet replied:
It creates a lambda that has no arguments.

IagoAbal commented on libs/ast_generic/AST_generic.ml:
```diff
@@ -626,6 +626,8 @@ and expr = {
   (* used to quickly get the range of an expression *)
   mutable e_range : (Tok.location * Tok.location) option;
       [@equal fun _a _b -> true] [@hash.ignore]
+  (* whether this expression is implicitly a return value in a function *)
```
Missing changelog.

amchiclet replied:
Oops, forgot to commit.

IagoAbal commented on src/analyzing/AST_to_IL.ml:
```diff
@@ -1047,7 +1047,11 @@ and stmt_expr env ?e_gen st =
+      (if eorig.is_implicit_return then
+       let ss, e = expr_with_pre_stmts ~void:true env eorig in
+       ss @ [ mk_s (Return (tok, e)) ] |> add_stmts env);
+      expr env eorig
```
Is it a good idea translating `eorig` twice? If `e` has side-effects, like a function call, this may be duplicating them. You could use `expr_with_pre_stmts` and if `not eorig.is_implicit_return` then you use `add_stmts ss` to put them back.

amchiclet replied:
Ah, yes, it should be translated once. Thanks.

IagoAbal commented on src/matching/Generic_vs_generic.ml:
```diff
@@ -2510,6 +2517,26 @@ and m_stmt a b =
   (* equivalence: *)
   | G.ExprStmt (a1, _), B.Return (_, Some b1, _) -> m_expr_deep a1 b1
+  (* implicit return *)
+  (* We try to match the FBExpr cases below where function
```
This stuff may fit better in `Matching_visitor`, when visiting an `ExprStmt` that is marked as "implicit return" then we could just generate the equivalence with a `return`, so nothing would change in `Generic_vs_generic`. But I don't have any strong opinion, the changes here are also not too ugly and they're concentrated in a single place. cc @aryx

amchiclet replied:
I'll try it out a bit to see if it actually fits better there or not.

amchiclet replied:
Good callout. Added some changes in `Matching_visitor`.

IagoAbal commented on src/naming/Naming_AST.ml:
```diff
@@ -471,6 +471,45 @@ let assign_implicitly_declares lang =
   lang =*= Lang.Python || lang =*= Lang.Ruby || lang =*= Lang.Php
   || Lang.is_js lang
 
+(*****************************************************************************)
```
I don't think this should go into `Naming`, they're unrelated. You can just add a new module in `analyzing/` that performs this step, and then perhaps call it from `Parse_target` when we already compute the CFG to perform const-prop.

amchiclet replied:
Sounds good.

amchiclet replied:
Moved things to a different file as suggested.

IagoAbal reviewed (changes requested):
See https://github.com/returntocorp/semgrep/pull/8826/files#r1338434051

amchiclet committed:
move analysis and matching to different places

amchiclet committed:
forgot to commit changelog

amchiclet committed:
add comments to mli file

amchiclet committed:
fix matching visitor super call

amchiclet commented:
Almost ready, but still requires a bit more updates.

Updating `Matching_visitor.ml` makes sense when visiting each root, so this part feels right to me.

But the change in `Generic_vs_generic.ml` is still needed when the user includes the context around `return $X`, such as trying to match

```
$FUN() = return $X
```

with

```
f() = 0
```

or matching

```
function $FUN()
  ...
  return $X
end
```

with

```
function f()
  0
end
```

which are examples mentioned in #8408.

amchiclet committed:
implicit return support in generic vs generic to support reported tests

amchiclet committed:
reorganize some functions

amchiclet committed:
added another test

amchiclet committed:
fix typo in comment

amchiclet committed:
fixed another typo

amchiclet commented:
This PR should be ready for another look. Addressed the first round of comments and added some more Julia tests to make sure it supports what was asked in #8408. This means the changes in `Generic_vs_generic` is still required, but the cases to match are simple because the changes in `Matching_visitor` are doing a lot of heavy lifting.

ihji commented on src/matching/Generic_vs_generic.ml:
```diff
@@ -1075,7 +1075,9 @@ and m_expr ?(is_root = false) ?(arguments_have_changed = true) a b =
   | G.Seq a1, B.Seq b1 -> (m_list m_expr) a1 b1
   | G.Ref (a0, a1), B.Ref (b0, b1) -> m_tok a0 b0 >>= fun () -> m_expr a1 b1
   | G.DeRef (a0, a1), B.DeRef (b0, b1) -> m_tok a0 b0 >>= fun () -> m_expr a1 b1
+  (* implicit return *)
```
shouldn't this be on the line just below?

amchiclet replied:
Yes, thanks!

ihji commented on src/analyzing/AST_to_IL.ml:
```diff
@@ -1231,55 +1236,55 @@ and mk_switch_break_label env tok =
+      let var = fresh_var env tok in
+      let lval = lval_of_base (Var var) in
+      let fake_i = mk_i (Assign (lval, e)) NoOrig in
+      [ mk_s (Instr fake_i) ]
```
Will we still require this placeholder node once we've removed the `implicit_return_hack`?

amchiclet replied:
Yes, we still need this placeholder, so it can be replaced with a return node. Without this placeholder, we wouldn't have a node to replace at all.

ihji commented on src/naming/dune:
```diff
@@ -11,6 +11,8 @@
    semgrep.core ; for Lang
    parser_javascript.ast ; ugly for Ast_js.default_entity
+   semgrep.il
+   pfff-lang_GENERIC-analyze
```
(Please excuse my lack of knowledge) What's the purpose of this `pfff-lang_GENERIC-analyze` dependency?

amchiclet replied:
Ooh! Thanks for catching that. It was needed because I needed to do CFG stuff inside the `naming` module. (The `analyze` module uses the `pfff-lang_GENERIC-analyze` name.)

But, now that the implementation has moved to the `analyze` module, I don't need this update in `naming` any more.

ihji reviewed (approved):
Looks great! Thanks for the awesome contribution.

amchiclet committed:
addressed some more review comments

IagoAbal commented on src/naming/Naming_AST.ml:
```diff
@@ -963,5 +1002,9 @@ let resolve lang prog =
   in
   visitor#visit_program () prog;
+
+  if lang_supports_implicit_return lang then
```
It's better to do this in `Parse_target` after `Constant_propagation.propagate_dataflow` and preferably sharing the CFG. It may require a bit of refactoring (but should be a small one). We could sync and talk about this.

amchiclet replied:
Somehow it seems like you're commenting on a previous commit, but the feedback applies, and also thanks for the offline discussion to clarify.

Made some updates. See more comments below.

amchiclet committed:
refactored cfg creation

amchiclet commented:
@IagoAbal : I experimented with your suggestions a bit and learned a couple of things.

> It's better to do this in Parse_target after Constant_propagation.propagate_dataflow

I tried this out and then it occurred to me that we don't really need to convert
```
    tmp = some_implicitly_returned_expr
```
to
```
    return some_implicitly_returned_expr
```
in the CFG.

We just need to mark `some_implicitly_returned_expr` so we can match it.

Originally I tried doing this implicit return analysis before constant propagation because I thought it may be helpful to have those return nodes, but it was irrelevant.

So with the most recent update, the Generic expressions are marked, but the IL nodes aren't converted to actual return nodes.

If there's some reason we prefer to actually have explicit return nodes when they are implicit, then I can also do another update, but this requires re-building the CFG(s).

> preferably sharing the CFG

Done. I've created a cfg for each function and iterate on the cfgs instead of creating them again.

IagoAbal commented on changelog.d/gh-8408.added:
```diff
@@ -0,0 +1,33 @@
+
+This matching can be disabled with the rule option `implicit_return`.
+
+For example:
```
I think you don't need this example. But we would need to update the docs.

amchiclet replied:
Ok I'll remove it.

IagoAbal reviewed (approved):
Could you rebase against develop? Before we merge, I want to run some benchmarks to check if adding that extra field to `expr` has any perf impact.

IagoAbal commented on src/analyzing/CFG_build.mli:
```diff
@@ -1 +1,13 @@
+ * For each function, returns a pair of the its parameters and its
+ * control flow graph.
+ *)
+val cfgs_in_program :
```
I was thinking more of a visitor rather than returning a list here, but if the perf or mem usage isn't affected (we'll measure it) then I don't mind this way. However, I think the "program itself as a top-level function" should be returned separately, I don't think you want to run the `Implicit_return` pass on that no?

Maybe:

```ocaml
type fun_cfg = { 
  fparams : IL.name list;
  fcfg : IL.cfg;
 }

type program_cfgs = fun_cfg list * fun_cfg

val cfgs_in_program : 
    Lang.t -> AST_generic.program -> program_cfgs
```
?

amchiclet replied:
Oh, interesting, I didn't think about a visitor. I'll return a list here (with your suggested changes to separate out the top-level function) but if benchmark results suggest otherwise, I can switch to the visitor.

IagoAbal commented on src/matching/Generic_vs_generic.ml:
```diff
@@ -2259,6 +2261,13 @@ and m_attribute a b =
+(*****************************************************************************)
+(* Implicit return *)
+(*****************************************************************************)
+and m_implicit_return (a : G.expr) (b : B.expr) tin =
```
Maybe add a comment here that `a` must come from a `return` statement?

amchiclet replied:
Done.

IagoAbal commented on src/parsing/Parse_target.ml:
```diff
@@ -97,6 +97,14 @@ let just_parse_with_lang lang file =
   | _else_ -> !just_parse_with_lang_ref lang file
   [@@profiling]
 
+let lang_supports_implicit_return (lang : Lang.t) =
```
This should probably be put in the `Implicit_return` module.

amchiclet replied:
Done.

IagoAbal commented on src/parsing/Parse_target.ml:
```diff
@@ -111,8 +119,23 @@ let parse_and_resolve_name lang file =
+
+  (* Flow-sensitive constant propagation. *)
+  cfgs
+  |> List.iter (fun (inputs, flow) ->
```
`Constant_propagation` and `Implicit_return` could just take a `program_cfgs` type and do the iteration, so here you would just have a call to `Constant_propagation.propagate_dataflow_progam cfgs` or something.

amchiclet replied:
Done.

IagoAbal commented on src/matching/Generic_vs_generic.ml:
```diff
@@ -2510,6 +2519,8 @@ and m_stmt a b =
   (* equivalence: *)
   | G.ExprStmt (a1, _), B.Return (_, Some b1, _) -> m_expr_deep a1 b1
+  (* implicit return *)
+  | G.Return (_, Some a1, _), B.ExprStmt (b1, _) -> m_implicit_return a1 b1
```
Is this still needed given the equivalence built into the `Matching_visitor` ?

IagoAbal replied:
I guess it is when there is a pattern like:
```
fun test() { 
  ...
  return $X
}
```
I believe you mentioned this before. Could we good to clarify in a comment perhaps attached to `m_implicit_return`.

amchiclet replied:
Yes, this is to handle the case when the pattern `return $X` is part of some surrounding code, e.g. the pattern
```
if x
  return 0
```
for the code
```
if x
  0
```

IagoAbal commented on src/matching/Matching_visitor.ml:
```diff
@@ -143,6 +144,17 @@ class ['self] matching_visitor =
+       * in Generic so we cannot match statement patterns against these
+       * expressions. This equivalence enables that.
+       *)
+      if env.implicit_return && e.is_implicit_return then
```
Also, is this needed once we have `m_implicit_return` in `Generic_vs_generic`? Would be good to clarify what cases do we match in each place and why are both needed.

amchiclet replied:
Yes, it's still needed. When we unconditionally add the equivalence here, we will be matching `return` vs `return` in `Generic_vs_generic`, which means `m_implicit_return` won't be called at all and this check would be bypassed in `Generic_vs_generic` if we don't have it here.

And true, it needs some clarification why. I've added more comments.

amchiclet committed:
address more review comments

amchiclet commented:
I've rebased on develop.

amchiclet commented:
Synced offline. Benchmarks look fine, so merging this.
